src/blktap.c

```diff
@@ -0,0 +1,328 @@
+/*
+ * Copyright (c) 2016 Assured Information Security, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+/**
+ * @file   blktap.c
+ * @author Jed Lejosne <lejosnej@ainfosec.com>
+ * @date   20 Oct 2016
+ *
+ * @brief  blktap-related functions
+ *
+ * This file implements all the code for handling the emulated CDROMs
+ * exposed to the guests. (the "iso" drive)
+ */
+
+#include "project.h"
+
+static int cdrom_vdev_of_domid(int domid)
+{
+  char xpath[256], **devs, *type;
+  unsigned int i, count;
+  int vdev, res = -1;
+
+  /* Find the CDROM virtual device for the given domid */
+  snprintf(xpath, sizeof(xpath), "/local/domain/0/backend/vbd/%d", domid);
+  devs = xs_directory(xs_handle, XBT_NULL, xpath, &count);
+  if (devs) {
+    for (i = 0; i < count; ++i) {
+      vdev = strtol(devs[i], NULL, 10);
+      type = xenstore_be_read(XBT_NULL, domid, vdev, "device-type");
```

Comment:
Hmm, I somehow managed to convince myself that the output of xs_read() didn't have to be free()ed! Fixing all the reads now...